### PR TITLE
Log the real reason when authentication fails (but don't show the user)

### LINKED ISSUE #24498: Failed authentication attempt is logged with text 'user does not exist' even when the user exists
### Description

**Log entry:**
...ers/web/auth/auth.go:206:SignInPost() [I] [645256b1] Failed authentication attempt for **_user_** from **_ip_**: user does not exist [uid: 2, name: **_user_**, keyid: 0]

**Description**
I tried to fail authentication by using wrong password and I looked at the log file. There is failed authentication attempt information log with additional text: 'user does not exist', however user exists and I can login when provided good password.

**Expected Fix**
Log authentication attempt fail log without text 'user does not exist' when user exists

### Gitea Version

1.19.2

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

2.30.2

### Operating System

Debian

### How are you running Gitea?

Running binary gitea behind nginx proxy

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by lunny**
Fix #24498

### CODE REVIEW

**Review by 6543 — APPROVED** [review]

File: services/auth/source/db/authenticate.go [hunk 1]
```diff
@@ -4,19 +4,66 @@
...
+type ErrUserPasswordInvalidate struct {
+	UID  int64
+	Name string
+}
+
+// IsErrUserPasswordInvalidate checks if an error is a ErrUserPasswordInvalidate
+func IsErrUserPasswordInvalidate(err error) bool {
+	_, ok := err.(ErrUserPasswordInvalidate)
+	return ok
+}
```
Comment (wxiaoguang) [inline services/auth/source/db/authenticate.go]: Why not use "errors.Is/As", but keeping adding these `IsErrXxx`?

File: routers/web/auth/auth.go [hunk 2]
```diff
@@ -201,7 +202,8 @@ func SignInPost(ctx *context.Context) {
 
 	u, source, err := auth_service.UserSignIn(form.UserName, form.Password)
 	if err != nil {
-		if user_model.IsErrUserNotExist(err) || user_model.IsErrEmailAddressNotExist(err) {
+		if user_model.IsErrUserNotExist(err) || user_model.IsErrEmailAddressNotExist(err) ||
+			auth_db.IsErrUserPasswordNotSet(err) || auth_db.IsErrUserPasswordInvalidate(err) {
```
Comment (wxiaoguang) [inline routers/web/auth/auth.go]: Could these errors be wrapped from / unwrapped to a generic auth error structure? Then no need to repeat these 4 `IsErrXxx` calls.

File: routers/web/auth/linkaccount.go [hunk 3]
```diff
@@ -81,6 +83,31 @@ func LinkAccount(ctx *context.Context) {
 	ctx.HTML(http.StatusOK, tplLinkAccount)
 }
 
+func handleSingInError(ctx *context.Context, userName string, ptrForm any, invoker string, err error) {
```
Comment (wxiaoguang) [inline routers/web/auth/linkaccount.go]: `SingIn` , sing ?

File: routers/web/auth/linkaccount.go [hunk 4]
```diff
@@ -81,6 +83,31 @@ func LinkAccount(ctx *context.Context) {
...
+	if user_model.IsErrUserNotExist(err) || user_model.IsErrEmailAddressNotExist(err) ||
+		auth_db.IsErrUserPasswordNotSet(err) || auth_db.IsErrUserPasswordInvalidate(err) {
+		ctx.Data["user_exists"] = true
```
Comment (wxiaoguang) [inline routers/web/auth/linkaccount.go:90]: `IsErrUserNotExist`, then how could `user_exists` be true?

File: routers/web/auth/linkaccount.go [hunk 5]
```diff
@@ -81,6 +83,31 @@ func LinkAccount(ctx *context.Context) {
...
+		ctx.RenderWithErr(ctx.Tr("form.username_password_incorrect"), tplConnectOID, ptrForm)
```
Comment (wxiaoguang) [inline routers/web/auth/linkaccount.go]: ~~Why the "linkaccount" could trigger "username_password_incorrect" ? No password is involved.~~

(misread)

File: services/auth/source/db/authenticate.go [hunk 6]
```diff
@@ -4,19 +4,66 @@
...
+
+func (err ErrUserPasswordInvalidate) Error() string {
+	return fmt.Sprintf("user's password is invalidated [uid: %d, name: %s]", err.UID, err.Name)
```
Comment (wxiaoguang) [inline services/auth/source/db/authenticate.go]: I think the message should be "is invalid"

`Invalidate` is a verb.

File: services/auth/source/db/authenticate.go [hunk 7]
```diff
@@ -4,19 +4,66 @@
...
+}
+
+// IsErrUserPasswordNotSet checks if an error is a ErrUserPasswordNotSet
+func IsErrUserPasswordNotSet(err error) bool {
+	_, ok := err.(ErrUserPasswordNotSet)
+	return ok
+}
+
+func (err ErrUserPasswordNotSet) Error() string {
+	return fmt.Sprintf("user's password doesn't set [uid: %d, name: %s]", err.UID, err.Name)
```
Comment (wxiaoguang) [inline services/auth/source/db/authenticate.go]: I think the message should be `isn't set` or `hasn't been set`.

**Review by wxiaoguang — APPROVED** [review]

File: services/auth/source/db/authenticate.go [hunk 8]
```diff
@@ -4,19 +4,54 @@
...
+	return fmt.Sprintf("user's password isn't set [uid: %d, name: %s]", err.UID, err.Name)
+}
+
+// Unwrap unwraps this error as a ErrInvalidArgument error
+func (err ErrUserPasswordNotSet) Unwrap() error {
+	return util.ErrInvalidArgument
+}
+
+// ErrUserPasswordInvalidate represents a "ErrUserPasswordInvalidate" kind of error.
+type ErrUserPasswordInvalidate struct {
```
Comment (delvh) [inline services/auth/source/db/authenticate.go]: I am not quite sure, but do you perhaps mean `ErrUserPasswordInvalid`?

File: routers/web/auth/auth.go [hunk 9]
```diff
@@ -201,7 +201,8 @@ func SignInPost(ctx *context.Context) {
 
 	u, source, err := auth_service.UserSignIn(form.UserName, form.Password)
 	if err != nil {
-		if user_model.IsErrUserNotExist(err) || user_model.IsErrEmailAddressNotExist(err) {
+		if user_model.IsErrUserNotExist(err) || user_model.IsErrEmailAddressNotExist(err) ||
```
Comment (delvh) [inline routers/web/auth/auth.go]: Can't we shorten that to
```suggestion
		if errors.Is(err, util.ErrNotExist) ||
```
?

File: routers/web/auth/linkaccount.go [hunk 10]
```diff
@@ -81,6 +83,32 @@ func LinkAccount(ctx *context.Context) {
 	ctx.HTML(http.StatusOK, tplLinkAccount)
 }
 
+func handleSignInError(ctx *context.Context, userName string, ptrForm any, tmpl base.TplName, invoker string, err error) {
+	if user_model.IsErrUserNotExist(err) || user_model.IsErrEmailAddressNotExist(err) {
```
Comment (delvh) [inline routers/web/auth/linkaccount.go]:
```suggestion
	if errors.Is(err, util.ErrNotExist) {
```

**Review by delvh — DISMISSED** [review]

File: routers/web/auth/linkaccount.go (hunk 10 above)
Comment (lunny) [reply routers/web/auth/linkaccount.go]: Done.

File: routers/web/auth/auth.go (hunk 9 above)
Comment (lunny) [reply routers/web/auth/auth.go]: Done.

**Comment by lunny** [comment]
@delvh done.

File: routers/web/auth/auth.go (hunk 9 above)
Comment (delvh) [reply routers/web/auth/auth.go]: Doesn't look like it yet here.

**Review by delvh — APPROVED** [review]
Apart from that below LGTM

**Pull request merged**